src/tools/upgrade-factory-snapshot.c

```diff
@@ -0,0 +1,725 @@
+            SFMF_WARN("Could not emit properties changed signal: %s\n", error->message);
+            g_error_free(error);
+        }
+
+        ufs->running = running;
+    }
+}
+
+void upgrade_factory_snapshot_dbus_signal_cb(GDBusConnection *connection,
+        const gchar *sender_name, const gchar *object_path, const gchar *interface_name,
+        const gchar *signal_name, GVariant *parameters, gpointer user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+
+    gchar *args = g_variant_print(parameters, TRUE);
+    //SFMF_DEBUG("dbus: sender=%s path=%s signal=%s.%s%s\n", sender_name, object_path,
+    //        interface_name, signal_name, args);
+    g_free(args);
+
+    if (strcmp(signal_name, "Progress") == 0) {
+        gchar *partition;
+        int progress;
+        gchar *message;
+        g_variant_get(parameters, "(sis)", &partition, &progress, &message);
+
+        upgrade_factory_snapshot_broadcast_status(ufs, partition, progress, message);
+
+        g_free(partition);
+        g_free(message);
+    } else {
+        SFMF_WARN("Unhandled D-Bus signal: '%s'\n", signal_name);
+    }
+}
+
+void upgrade_factory_snapshot_method_call_cb(GDBusConnection *connection, const gchar *sender,
+        const gchar *object_path, const gchar *interface_name, const gchar *method_name,
+        GVariant *parameters, GDBusMethodInvocation *invocation, gpointer user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+    GVariant *return_value = NULL;
+
+    if (g_strcmp0(object_path, UFS_DBUS_PATH) == 0 &&
+            g_strcmp0(interface_name, UFS_DBUS_INTERFACE) == 0 &&
+            sfmf_dbus_is_privileged(connection, sender)) {
+        if (g_strcmp0(method_name, "Start") == 0) {
+            gboolean result = FALSE;
+
+            if (!ufs->running) {
+                gchar *release = 0;
+                g_variant_get(parameters, "(s)", &release);
+                if (release && strlen(release)) {
+                    setenv("SSU_SLIPSTREAM_RELEASE", release, 1);
+                }
+                g_free(release);
+
+                g_idle_add(upgrade_factory_snapshot_do_next_entry_cb, ufs);
+                upgrade_factory_snapshot_set_running(ufs, TRUE);
+                result = TRUE;
+            }
+
+            return_value = g_variant_new("(b)", result);
+        } else if (g_strcmp0(method_name, "GetProgress") == 0) {
+            struct DeployTaskQueue *queue = ufs->deploy_queue;
+            struct DeployTask *task = deploy_task_queue_get_current_task(queue);
+
+            return_value = g_variant_new("(ssiisiisi)", queue->name,
+                    task ? task->name : "", queue->current+1, queue->total,
+                    ufs->status.partition ?: "", ufs->status.partition_current+1, ufs->status.partition_total,
+                    ufs->status.message ?: "", ufs->status.progress);
+        }
+    }
+
+    if (return_value) {
+        g_dbus_method_invocation_return_value(invocation, return_value);
```

Comment:
Yes, from [the docs](https://developer.gnome.org/gio/stable/GDBusMethodInvocation.html#g-dbus-method-invocation-return-value):

"This method will free invocation, you cannot use it afterwards."